cmd/agent/workspace/setup_gpg.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/loft-sh/log"
+	"github.com/spf13/cobra"
+)
+
+// SetupGPGCmd holds the setupGPG cmd flags
+type SetupGPGCmd struct {
+	*flags.GlobalFlags
+
+	PublicKey  string
+	OwnerTrust string
+	SocketPath string
+}
+
+// NewSetupGPGCmd creates a new command
+func NewSetupGPGCmd(flags *flags.GlobalFlags) *cobra.Command {
+	cmd := &SetupGPGCmd{
+		GlobalFlags: flags,
+	}
+	setupGPGCmd := &cobra.Command{
+		Use:   "setup-gpg",
+		Short: "setups gpg-agent forwarding in the container",
+		Args:  cobra.NoArgs,
+		RunE: func(_ *cobra.Command, _ []string) error {
+			return cmd.Run(context.Background())
+		},
+	}
+	setupGPGCmd.Flags().StringVar(&cmd.PublicKey, "publickey", "", "GPG Public keys to import in armor form")
+	setupGPGCmd.Flags().StringVar(&cmd.OwnerTrust, "ownertrust", "", "GPG Owner trust to import in armor form")
+	setupGPGCmd.Flags().StringVar(&cmd.SocketPath, "socketpath", "", "patht to the gpg socket forwarded")
+	return setupGPGCmd
+}
+
+// will forward a local gpg-agent into the remote container
+// this works by
+//
+// - stopping remote gpg-agent and removing the sockets
+// - exporting local public keys and owner trust
+// - importing those into the container
+// - ensuring the gpg-agent is stopped in the container
+// - starting a reverse-tunnel of the local unix socket to remote
+// - ensuring paths and permissions are correctly set in the remote
+func (cmd *SetupGPGCmd) Run(ctx context.Context) error {
+	logger := log.Default.ErrorStreamOnly()
+
+	logger.Debugf("Initializing gpg-agent forwarding")
+
+	logger.Debugf("Stopping container gpg-agent")
+	err := cmd.stopGpgAgent()
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Decoding input public key")
+	publicKey, err := base64.StdEncoding.DecodeString(cmd.PublicKey)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("gpg: importing gpg public key in container")
+	err = cmd.importGpgKey(publicKey)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Decoding input owner trust")
+	ownerTrust, err := base64.StdEncoding.DecodeString(cmd.OwnerTrust)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("gpg: importing gpg owner trust in container")
+	err = cmd.importOwnerTrust(ownerTrust)
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Ensuring paths existence and permissions")
+	err = cmd.setupRemoteSocketDirtree()
+	if err != nil {
+		return err
+	}
+
+	// Now we again kill the agent and remove the socket to really be sure every
+	// thing is clean
+	logger.Debugf("Ensure stopping container gpg-agent")
+	err = cmd.stopGpgAgent()
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Setup local gnupg dirs")
+	err = cmd.setupLocalGpg()
+	if err != nil {
+		return err
+	}
+
+	logger.Debugf("Setup gpg.conf")
+	err = cmd.setupGpgConf()
+	if err != nil {
+		return err
+	}
+
+	time.Sleep(time.Second)
+
+	return nil
+}
+
+func (cmd *SetupGPGCmd) stopGpgAgent() error {
+	return exec.Command("gpgconf", []string{"--kill", "gpg-agent"}...).Run()
```

Comment:
does this fail if the agent is already stopped?

Reply:
Nope and that's ok we want to have it stopped anyway